src/i_winmusic.c

```diff
@@ -176,93 +165,116 @@ static void StreamOut(void)
-    }
+    DWORD events[3] = {0};
+    events[0] = 0;                                             // dwDeltaTime
+    events[1] = 0;                                             // dwStreamID
+    events[2] = MAKE_EVT(type, param1, param2, MEVT_SHORTMSG); // dwEvent
+    WriteBuffer((byte *)events, 3 * sizeof(DWORD));
 }
 
-// The Windows API documentation states: "Applications should not call any
-// multimedia functions from inside the callback function, as doing so can
-// cause a deadlock." We use thread to avoid possible deadlocks.
+static void SendLongMsg(const byte *ptr, int length)
+{
+    DWORD events[3] = {0};
+    events[0] = 0;
+    events[1] = 0;
+    events[2] = MAKE_EVT(length, 0, 0, MEVT_LONGMSG);
+    WriteBuffer((byte *)events, 3 * sizeof(DWORD));
+    WriteBuffer(ptr, length);
+}
 
-static DWORD WINAPI PlayerProc(void)
+static void ResetDevice(void)
 {
-    HANDLE events[2] = { hBufferReturnEvent, hExitEvent };
+    int i;
 
-    while (1)
+    for (i = 0; i < MIDI_CHANNELS_PER_TRACK; ++i)
     {
-        switch (WaitForMultipleObjects(2, events, FALSE, INFINITE))
-        {
-            case WAIT_OBJECT_0:
-                FillBuffer();
-                StreamOut();
-                break;
+        // RPN sequence to adjust pitch bend range (RPN value 0x0000)
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x65, 0x00);
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x64, 0x00);
 
-            case WAIT_OBJECT_0 + 1:
-                return 0;
-        }
-    }
-    return 0;
-}
+        // reset pitch bend range to central tuning +/- 2 semitones and 0 cents
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x06, 0x02);
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x26, 0x00);
 
-static void FreeSong(void)
-{
-    if (song.native_events)
-    {
-        free(song.native_events);
-        song.native_events = NULL;
+        // end of RPN sequence
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x64, 0x7f);
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x65, 0x7f);
+
+        // reset all controllers
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x79, 0x00);
+
+        // reset pan to 64 (center)
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x0a, 0x40);
+
+        // reset reverb to 40 and other effect controllers to 0
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5b, winmm_reverb_level); // reverb
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5c, 0x00); // tremolo
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5d, winmm_chorus_level); // chorus
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5e, 0x00); // detune
+        SendShortMsg(MIDI_EVENT_CONTROLLER | i, 0x5f, 0x00); // phaser
     }
-    song.num_events = 0;
-    song.position = 0;
-    song.looping = false;
 }
 
-// Convert a multi-track MIDI file to an array of Windows MIDIEVENT structures.
+static byte gs_reset[] = {0xf0, 0x7e, 0x7f, 0x09, 0x01, 0xf7};
+static byte master_vol[] = {0xf0, 0x7f, 0x7f, 0x04, 0x01, 0x7f, 0x7f, 0xf7};
```

Comment:
I take this from [ZMusic](https://github.com/ZDoom/ZMusic/blob/1dda85bbe1ae2c272a57566fb8698ba7d660d51f/source/musicformats/music_midi.cpp#L817), but I noticed that in PrBoom+ these messages are different.

Reply:
1. The comment in ZMusic says GS Reset but the message is actually GM System On.
2. `master_vol[5`] and `master_vol[6]` are where the volume value is stored, so `0x7F` `0x7F` is full volume (16383). It's different in [PrBoom+](https://github.com/coelckers/prboom-plus/blob/master/prboom2/src/MUSIC/portmidiplayer.c#L452), because those values are ignored in the comparison that's being done.

Reply:
Fixed in https://github.com/fabiangreffrath/woof/pull/816/commits/28fd74ee26fca94d0ab1f642543516f418f5844b I wonder if GS Reset is enough, or are other types of resets also needed?